lib/internal/process/promises.js

```diff
@@ -7,13 +7,52 @@ const pendingUnhandledRejections = [];
 const asyncHandledRejections = [];
 let lastPromiseId = 0;
 
+const kUnhandledRejection = 0;
+const kHandledRejection = 1;
+const kResolveAfterResolved = 2;
+const kRejectAfterResolved = 3;
+
 exports.setup = setupPromises;
 
 function setupPromises(_setupPromises) {
-  _setupPromises(unhandledRejection, handledRejection);
+  _setupPromises(handler);
   return emitPromiseRejectionWarnings;
 }
 
+function handler(type, promise, reason) {
+  switch (type) {
+    case kUnhandledRejection:
+      return unhandledRejection(promise, reason);
+    case kHandledRejection:
+      return handledRejection(promise);
+    case kResolveAfterResolved:
+      return resolveError('resolveAfterResolved', promise, reason,
+                          'Resolve was called more than once.');
+    case kRejectAfterResolved:
+      return resolveError('rejectAfterResolved', promise, reason,
+                          'Reject was called after resolve.');
+  }
+  // eslint-disable-next-line no-restricted-syntax
+  throw new Error('Unreachable');
+}
+
+let warned = false;
```

Comment:
nit: should name this something like `resolveAfterResolveWarned` or something more specific just in case we need to add multiple such flags at some point